src/event_message/mod.rs

```diff
@@ -141,7 +141,30 @@ impl EventSemantics for EventMessage {
 
 impl EventSemantics for SignedEventMessage {
     fn apply_to(&self, state: IdentifierState) -> Result<IdentifierState, Error> {
-        self.event_message.apply_to(state)
+        // Update state.last with serialized current event message.
+        let new_state = match &self.event_message.event.event_data {
+            EventData::Icp(_) => IdentifierState {
+                last: self.serialize()?,
+                ..state
+            },
+            EventData::Rot(rot) => {
+                // Check if hashes of state.last event and previous_event_hash matches.
+                if rot.previous_event_hash.derivation.derive(&state.last) == rot.previous_event_hash
+                {
+                    IdentifierState {
+                        last: self.serialize()?,
+                        ..state
+                    }
+                } else {
+                    return Err(Error::SemanticError(
+                        "Last event does not match previous event".to_string(),
+                    ));
+                }
+            }
+            _ => todo!(),
+        };
+
+        self.event_message.apply_to(new_state)
```

Comment:
this logic all looks correct, however the hash is made with the serialized previous event without appended signatures (`last` should be a serialized `EventMessage`), so this can be moved almost as-is to `EventMessage::apply_to` and it should be sweet :+1: